### PR TITLE
Add gRPC-OpenTracing interceptors for Go

### PR DESCRIPTION
**Pull request by bhs**
See the `README.md` for usage examples, etc.

This is possible thanks to https://github.com/grpc/grpc-go/pull/867.

### CODE REVIEW

**Comment by bhs** [comment]
cc: @dominikh per our conversation on Gitter about this

I would like to include more information about the peer (via tags/etc), but unfortunately I'm not sure where to access that given what's (not) exported in gRPC.

**Comment by bhs** [comment]
Per the google CLA thing: "I did it"

**Comment by bg451** [comment]
Everything looks good to me, although I would recommend adding an `Option` that allows users to specify a set of tags that'll be applied whenever a new span is created in the interceptor.

File: go/otgrpc/client.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+			ext.SpanKindRPCClient,
+			gRPCComponentTag,
+		)
+		defer clientSpan.Finish()
+		md, ok := metadata.FromContext(ctx)
+		if !ok {
+			md = metadata.New(nil)
+		}
+		mdWriter := metadataReaderWriter{md}
+		tracer.Inject(clientSpan.Context(), opentracing.TextMap, mdWriter)
```
Comment (jmacd) [inline go/otgrpc/client.go]: There's an error to handle, or a comment explaining why not.

**Comment by jmacd** [comment]
👍

File: go/otgrpc/client.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+			gRPCComponentTag,
+		)
+		defer clientSpan.Finish()
+		md, ok := metadata.FromContext(ctx)
+		if !ok {
+			md = metadata.New(nil)
+		}
+		mdWriter := metadataReaderWriter{md}
+		err = tracer.Inject(clientSpan.Context(), opentracing.TextMap, mdWriter)
+		// We have no better place to record an error than the Span itself :-/
```
Comment (bhs) [inline go/otgrpc/client.go:56]: @jmacd this is how I dealt with Inject errors...

image

It gets the job done. Registering an error handler as an `Option` seemed like overkill.

File: go/otgrpc/options.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+// arbitrary tags/logs/etc to the opentracing.Span associated with client
+// and/or server RPCs.
+type SpanDecoratorFunc func(
+	span opentracing.Span,
+	method string,
+	req, resp interface{},
+	grpcError error)
+
+// SpanDecorator binds a function that decorates gRPC Spans.
+func SpanDecorator(decorator SpanDecoratorFunc) Option {
```
Comment (bhs) [inline go/otgrpc/options.go:30]: @bg451 this makes it easy to decorate client and/or server Spans with arbitrary data, including things gleaned from app-level requests and responses.

**Comment by bhs** [comment]
Comments addressed... merging.

**Pull request merged**